engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/AliasAnalysis.scala

```diff
@@ -656,7 +656,7 @@ case object AliasAnalysis extends IRPass {
             )
           )
           .updateMetadata(this -->> Info.Scope.Child(graph, currentScope))
-      case _: Function.Binding =>
+      case _ =>
         throw new CompilerError(
           "Function sugar should not be present during alias analysis."
```

Comment:
It's not necessarily a `Function.Binding` so probably worth adding a class of the value to the error at least

Reply:
Advice from Scala expert is needed: Why the Scala compiler complains the pattern isn't exhaustive now and not before this PR?

Reply:
Fixed for you in https://github.com/enso-org/enso/pull/8207/commits/0fec07df736cd37335cad35ec695a85f755e020f. Feel free to revert the rest of the changed patterns. Now they will pass without warnings.

This was due to a custom `unapply` that could return an `Option` rather than `Some` and therefore confuse the exhaustiveness checker. Since it takes `Lambda` we know that it will always return the latter.

Reply:
I cannot get the other cases to work. Adding following changes on top of [385b3b8](https://github.com/enso-org/enso/pull/8207/commits/385b3b800d31585fb3ec9dd35e96b41ea0a400b7) again fails with non-exhaustive pattern error:
```diff
enso$ git diff
diff --git engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/DataflowAnalysis.scala engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/DataflowAnalysis.scala
index 0a5f58b8b8..5c0e7695aa 100644
--- engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/DataflowAnalysis.scala
+++ engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/DataflowAnalysis.scala
@@ -308,7 +308,7 @@ case object DataflowAnalysis extends IRPass {
             body      = analyseExpression(body, info)
           )
           .updateMetadata(this -->> info)
-      case _ =>
+      case _: Function.Lambda =>
         throw new CompilerError(
           "Function sugar should not be present during dataflow analysis."
         )
diff --git engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/DemandAnalysis.scala engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/DemandAnalysis.scala
index 8be2b2c97c..b7619031ec 100644
--- engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/DemandAnalysis.scala
+++ engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/DemandAnalysis.scala
@@ -160,7 +160,7 @@ case object DemandAnalysis extends IRPass {
             isInsideCallArgument = false
           )
         )
-      case _ =>
+      case _: Function.Lambda =>
         throw new CompilerError(
           "Function sugar should not be present during demand analysis."
         )
diff --git engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/TailCall.scala engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/TailCall.scala
index 48c00fe723..e0af481a45 100644
--- engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/TailCall.scala
+++ engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/TailCall.scala
@@ -450,7 +450,7 @@ case object TailCall extends IRPass {
           arguments = args.map(analyseDefArgument),
           body      = analyseExpression(body, isInTailPosition = markAsTail)
         )
-      case _ =>
+      case _: Function.Lambda =>
         throw new CompilerError(
           "Function sugar should not be present during tail call analysis."
         )
```

Reply:
idk. I added them without any issues.

Reply:
OK. Thank you.